File: src/test/java/edu/metrostate/trackingsystem/Scene.java
package edu.metrostate.trackingsystem;

import edu.metrostate.trackingsystem.domain.models.Dealer;
import edu.metrostate.trackingsystem.domain.models.Price;
import edu.metrostate.trackingsystem.domain.models.Vehicle;

import java.time.Instant;
import java.util.List;
import java.util.Objects;

import static org.junit.jupiter.api.Assertions.*;

public class Scene {

    private final Main.DependencyPackage dp = Main.getDependencies(null);

    private int lastDealerIndex = 0;
    private int lastVehicleIndex = 0;

    public Scene setup() {
        dp.database().getDealers().clear();
        return this;
    }

    public Scene dealer(String dealerName, boolean enabledForAcquisition) {
        var dealer = new Dealer("#" + ++lastDealerIndex, dealerName);
        if (!enabledForAcquisition) {
            dealer.setEnabledForAcquisition(false);
        }
        dp.database().getDealers().add(dealer);
        return this;
    }

    /**
     * @param vehicle dealer#1 ford f-150 pickup 50000 2/3/2025
     */
    public Scene vehicle(String vehicle) {
        addVehicle(vehicle);
        return this;
    }

    public boolean addVehicle(String vehicle) {
        var vehicleId = "#" + ++lastVehicleIndex;

        var parts = vehicle.split(" ");
        var dealershipId = parts[0].replace("dealer", "");
        var manufacturer = parts[1];
        var model = parts[2];
        var type = parts[3];
        var price = Double.parseDouble(parts[4]);
        long acquisitionDate = Instant.now().toEpochMilli();

        var newVehicle = new Vehicle(vehicleId, manufacturer, model, acquisitionDate, new Price(price, "dollars"), dealershipId, type);
        return dp.vehicleService().addVehicle(newVehicle).isSuccess();
    }

    public boolean deleteVehicle(String vehicleId) {
        var effectiveId = vehicleId.replace("vehicle", "");

        var vehicle = dp.vehicleService()
                .getVehicles()
                .stream()
                .filter(v -> Objects.equals(v.getVehicleId(), effectiveId)) // Filter the matching vehicle
                .findFirst()
                .orElse(null);

        assertNotNull(vehicle);
        return dp.vehicleService().deleteVehicle(effectiveId, vehicle.getDealershipId()).isSuccess();
    }

    public List<Dealer> getDealers() {
        return dp.dealerService().getDealers();
    }

    public Dealer getDealer(String dealershipId) {
        return dp.database().getDealerByID(dealershipId);
    }
}

File: src/test/kotlin/edu/metrostate/dealership/Scene.kt
package edu.metrostate.dealership

import edu.metrostate.dealership.domain.models.Dealer
import edu.metrostate.dealership.domain.models.Price
import edu.metrostate.dealership.domain.models.Vehicle
import edu.metrostate.dealership.domain.models.VehicleType.Companion.fromString
import java.io.File
import java.time.Instant
import java.util.*

class Scene {
    private val dp = Main.Companion.getDependencies()
    fun setup(): Scene {
        dp.database.dealers.clear()
        dp.database.vehicles.clear()
        vehicleDealerMap.clear()
        lastVehicleIndex = 0
        lastDealerIndex = 0
        return this
    }

    fun dealer(dealerName: String?, enabledForAcquisition: Boolean): Scene {
        val dealer = Dealer("#" + ++lastDealerIndex, dealerName ?: "Unknown", true)
        if (!enabledForAcquisition) {
            dealer.enabledForAcquisition = false
        }
        dp.database.dealers.add(dealer)
        return this
    }

    /**
     * @param vehicle dealer#1 ford f-150 pickup 50000 2/3/2025
     */
    fun vehicle(vehicle: String): Scene {
        addVehicle(vehicle)
        return this
    }

    fun addVehicle(vehicle: String, overrideLastVehicleIndex: Int): Boolean {
        lastVehicleIndex = overrideLastVehicleIndex
        return addVehicle(vehicle)
    }

    fun addVehicle(vehicle: String): Boolean {
        val vehicleId = "#" + ++lastVehicleIndex

        val parts = vehicle.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val dealershipId = parts[0].replace("dealer", "")
        val manufacturer = parts[1]
        val model = parts[2]
        val type = parts[3]
        val price = parts[4].toDouble()
        val acquisitionDate = Instant.now().toEpochMilli()

        val newVehicle = Vehicle(
            vehicleId,
            manufacturer,
            model,
            acquisitionDate,
            Price(price, "dollars"),
            dealershipId,
            fromString(type),
            false
        )
        val isSuccess = dp.vehicleService.addVehicle(newVehicle).isSuccess
        if (isSuccess) {
            vehicleDealerMap[vehicleId] = dealershipId
        }
        return isSuccess
    }

    fun deleteVehicle(vehicleId: String): Boolean {
        val effectiveVehicleId = vehicleId.replace("vehicle", "")
        return dp.vehicleService.deleteVehicle(effectiveVehicleId)
    }

    val dealers: List<Dealer>
        get() = dp.dealerService.dealers

    val vehicles: List<Vehicle>
        get() = dp.vehicleService.vehicles

    fun getDealer(dealershipId: String): Dealer? {
        return dp.database.getDealerByID(dealershipId)
    }

    fun importJson(): Boolean {
        val file = File("src/test/kotlin/jsonTest.json")
        return dp.jsonHandler.importFile(file)
    }

    fun importXml(): Boolean {
        val file = File("src/test/kotlin/xmlTest.xml")
        return dp.xmlHandler.importFile(file)
    }

    companion object {
        private val vehicleDealerMap = HashMap<String, String>()

        private var lastDealerIndex = 0
        private var lastVehicleIndex = 0
    }
}